### ISSUE TITLE
[cantera 2.3.0] surfSolverTest/surfaceSolver.cpp not compiled

### ISSUE THREAD

**Issue by sagitter** [Cantera/cantera#349]
Hi.

surfSolverTest is not compiled on Fedora 24 (fmt-3.0.0):

```
g++ -o test_problems/surfSolverTest/surfaceSolver.o -c -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic --std=c++14 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -pthread -DNDEBUG -DNDEBUG -Iinclude -Isrc -Itest_problems/shared -I/usr/include/eigen3 -I/usr/include/boost -I/usr/include test_problems/surfSolverTest/surfaceSolver.cpp
test_problems/surfSolverTest/surfaceSolver.cpp: In function 'void printGas(std::ostream&, Cantera::ThermoPhase*, Cantera::InterfaceKinetics*, double*)':
test_problems/surfSolverTest/surfaceSolver.cpp:60:69: error: no matching function for call to 'print(std::ostream&, const char [45], size_t&, std::__cxx11::string, double&, double&, double&)'
                    k, gasTP->speciesName(k), C[k], x[k], src[kstart]);
                                                                     ^
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3137:7: note: candidate: void fmt::print(FILE*, fmt::CStringRef, fmt::ArgList)
 FMT_API void print(std::FILE *f, CStringRef format_str, ArgList args);
       ^~~~~
/usr/include/fmt/format.h:3137:7: note:   candidate expects 3 arguments, 7 provided
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3148:7: note: candidate: void fmt::print(fmt::CStringRef, fmt::ArgList)
 FMT_API void print(CStringRef format_str, ArgList args);
       ^~~~~
/usr/include/fmt/format.h:3148:7: note:   candidate expects 2 arguments, 7 provided
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3463:34: note: candidate: template<class ... Args> void fmt::print(fmt::CStringRef, const Args& ...)
 FMT_VARIADIC(void, print, CStringRef)
                                  ^~~~ 
/usr/include/fmt/format.h:3463:34: note:   template argument deduction/substitution failed:
test_problems/surfSolverTest/surfaceSolver.cpp:59:20: note:   cannot convert 'oooo' (type 'std::ostream {aka std::basic_ostream<char>}') to type 'fmt::CStringRef {aka fmt::BasicCStringRef<char>}'
         fmt::print(oooo, "{:4d} {:>24s}   {:14.3g} {:14.3g}  {:14.3e}\n",
                    ^~~~
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3464:34: note: candidate: template<class ... Args> void fmt::print(FILE*, fmt::CStringRef, const Args& ...)
 FMT_VARIADIC(void, print, std::FILE *, CStringRef)
                                  ^~~~~
/usr/include/fmt/format.h:3464:34: note:   template argument deduction/substitution failed:
test_problems/surfSolverTest/surfaceSolver.cpp:59:20: note:   cannot convert 'oooo' (type 'std::ostream {aka std::basic_ostream<char>}') to type 'FILE* {aka _IO_FILE*}'
         fmt::print(oooo, "{:4d} {:>24s}   {:14.3g} {:14.3g}  {:14.3e}\n",
                    ^~~~
test_problems/surfSolverTest/surfaceSolver.cpp: In function 'void printBulk(std::ostream&, Cantera::ThermoPhase*, Cantera::InterfaceKinetics*, double*)':
test_problems/surfSolverTest/surfaceSolver.cpp:95:75: error: no matching function for call to 'print(std::ostream&, const char [45], size_t&, std::__cxx11::string, double&, double&, double&)'
                    k, bulkPhaseTP->speciesName(k), C[k], x[k], src[kstart]);
                                                                           ^
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3137:7: note: candidate: void fmt::print(FILE*, fmt::CStringRef, fmt::ArgList)
 FMT_API void print(std::FILE *f, CStringRef format_str, ArgList args);
       ^~~~~
/usr/include/fmt/format.h:3137:7: note:   candidate expects 3 arguments, 7 provided
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3148:7: note: candidate: void fmt::print(fmt::CStringRef, fmt::ArgList)
 FMT_API void print(CStringRef format_str, ArgList args);
       ^~~~~
/usr/include/fmt/format.h:3148:7: note:   candidate expects 2 arguments, 7 provided
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3463:34: note: candidate: template<class ... Args> void fmt::print(fmt::CStringRef, const Args& ...)
 FMT_VARIADIC(void, print, CStringRef)
                                  ^~~~ 
/usr/include/fmt/format.h:3463:34: note:   template argument deduction/substitution failed:
test_problems/surfSolverTest/surfaceSolver.cpp:94:20: note:   cannot convert 'oooo' (type 'std::ostream {aka std::basic_ostream<char>}') to type 'fmt::CStringRef {aka fmt::BasicCStringRef<char>}'
         fmt::print(oooo, "{:4d} {:>24s}   {:14.3g} {:14.3g}  {:14.3e}\n",
                    ^~~~
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3464:34: note: candidate: template<class ... Args> void fmt::print(FILE*, fmt::CStringRef, const Args& ...)
 FMT_VARIADIC(void, print, std::FILE *, CStringRef)
                                  ^~~~~
/usr/include/fmt/format.h:3464:34: note:   template argument deduction/substitution failed:
test_problems/surfSolverTest/surfaceSolver.cpp:94:20: note:   cannot convert 'oooo' (type 'std::ostream {aka std::basic_ostream<char>}') to type 'FILE* {aka _IO_FILE*}'
         fmt::print(oooo, "{:4d} {:>24s}   {:14.3g} {:14.3g}  {:14.3e}\n",
                    ^~~~
test_problems/surfSolverTest/surfaceSolver.cpp: In function 'void printSurf(std::ostream&, Cantera::ThermoPhase*, Cantera::InterfaceKinetics*, double*)':
test_problems/surfSolverTest/surfaceSolver.cpp:136:62: error: no matching function for call to 'print(std::ostream&, const char [37], size_t&, std::__cxx11::string, double&, double&)'
                    k, surfPhaseTP->speciesName(k), x[k], srcK);
                                                              ^
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3137:7: note: candidate: void fmt::print(FILE*, fmt::CStringRef, fmt::ArgList)
 FMT_API void print(std::FILE *f, CStringRef format_str, ArgList args);
       ^~~~~
/usr/include/fmt/format.h:3137:7: note:   candidate expects 3 arguments, 6 provided
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3148:7: note: candidate: void fmt::print(fmt::CStringRef, fmt::ArgList)
 FMT_API void print(CStringRef format_str, ArgList args);
       ^~~~~
/usr/include/fmt/format.h:3148:7: note:   candidate expects 2 arguments, 6 provided
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3463:34: note: candidate: template<class ... Args> void fmt::print(fmt::CStringRef, const Args& ...)
 FMT_VARIADIC(void, print, CStringRef)
                                  ^~~~ 
/usr/include/fmt/format.h:3463:34: note:   template argument deduction/substitution failed:
test_problems/surfSolverTest/surfaceSolver.cpp:135:20: note:   cannot convert 'oooo' (type 'std::ostream {aka std::basic_ostream<char>}') to type 'fmt::CStringRef {aka fmt::BasicCStringRef<char>}'
         fmt::print(oooo, "{:4d} {:>24s}   {:14.3g}   {:14.3e}\n",
                    ^~~~
In file included from include/cantera/base/ctexceptions.h:14:0,
                 from include/cantera/thermo/Phase.h:10,
                 from include/cantera/thermo/ThermoPhase.h:13,
                 from include/cantera/thermo.h:12,
                 from include/cantera/Interface.h:8,
                 from test_problems/surfSolverTest/surfaceSolver.cpp:17:
/usr/include/fmt/format.h:3464:34: note: candidate: template<class ... Args> void fmt::print(FILE*, fmt::CStringRef, const Args& ...)
 FMT_VARIADIC(void, print, std::FILE *, CStringRef)
                                  ^~~~~
/usr/include/fmt/format.h:3464:34: note:   template argument deduction/substitution failed:
test_problems/surfSolverTest/surfaceSolver.cpp:135:20: note:   cannot convert 'oooo' (type 'std::ostream {aka std::basic_ostream<char>}') to type 'FILE* {aka _IO_FILE*}'
         fmt::print(oooo, "{:4d} {:>24s}   {:14.3g}   {:14.3e}\n",
                    ^~~~
scons: *** [test_problems/surfSolverTest/surfaceSolver.o] Error 1
scons: building terminated because of errors.
error: Bad exit status from /var/tmp/rpm-tmp.25pz9b (%check)
```

**Comment by speth** [comment]
The only version of fmt/cppformat I can find for FC24 is 2.1.1. How are you getting 3.0?

**Comment by sagitter** [comment]
> The only version of fmt/cppformat I can find for FC24 is 2.1.1. How are you getting 3.0?

Sorry, I had forgotten to mention that `cppformat` will be soon replaced by `fmt-3.0.0` ([bugzilla ticket](https://bugzilla.redhat.com/show_bug.cgi?id=1350143)). If you need `fmt`for testing use [these packages](http://copr-fe.cloud.fedoraproject.org/coprs/sagitter/petsc/build/368285/) please.

**Comment by speth** [comment]
The packages you linked to are missing `ostream.h` which should be present in `fmt` version 3.0.0.

**Comment by sagitter** [comment]
Yes, I seen. I don't know why original packager has excluded them, i will investigate.
i'm rebuilding `fmt`.

**Comment by sagitter** [comment]
I'm compiling https://github.com/Cantera/cantera/commit/97e84aaf9b0da13480bdf610d12b7d792acdbf74 with  `system_fmt=y`:

```
Copy("build/data/elements.xml", "data/inputs/elements.xml")
scons: *** [build/ext/LICENSE.txt] Source `ext/fmt/LICENSE.rst' not found, needed by target `build/ext/LICENSE.txt'.

```

**Comment by speth** [comment]
Sorry, that should be fixed in 72d8bb0.

**Comment by sagitter** [comment]
Fixed; thank you.
